### PR TITLE
1218 check delegate wallets

### LINKED ISSUE #1218: Check delegate wallets
Add check that operational wallet and delegate wallet differs. If they are the same for providers, reject this provider during the registration. Related to all provider types: miner, sharder, blobber, authorizer, validator.

### PR DESCRIPTION
**Pull request by devyetii**
## Fixes
- #1218
- Delegate wallets and operational wallet of a provider should differ
## Changes
- Add checks for this condition in the registration phase of the providers in the corresponding smart contracts.

## Need to be mentioned in CHANGELOG.md?
- Yes

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by devyetii** [comment]
@dabasov Kindly check

**Comment by peterlimg** [comment]
@devyetii is this still WIP, or is ready to review

**Comment by devyetii** [comment]
@peterlimg waiting for review from @dabasov

File: code/go/0chain.net/smartcontract/minersc/miner.go
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/miner.go]: Let's avoid capital error message

File: code/go/0chain.net/smartcontract/minersc/miner.go
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/miner.go]: Avoid capital error message, and use consistent message, so change "Can't" to "couldn't" or vice verse. Also I saw repeat code below that basically do the same thing, which we can improve by moving this code to a function, and define in parent package `code/go/0chain.net/smartcontract`.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/minersc/miner.go
Comment (devyetii) [reply code/go/0chain.net/smartcontract/minersc/miner.go]: Done

File: code/go/0chain.net/smartcontract/minersc/miner.go
Comment (devyetii) [reply code/go/0chain.net/smartcontract/minersc/miner.go]: Done

File: code/go/0chain.net/smartcontract/common/wallet.go
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/common/wallet.go:28]: Oh, actually this functions will be reused somewhere else and should be put in the `encryption` package, where we do the public/private/client ID things. I've done that in another [PR](https://github.com/0chain/0chain/pull/1850/files#diff-a4b5842743d9fb9c0c1e8f9d21e4b238ccf11fa0cb59fef25a8890ce57180622R176-R188) but not merged it, let me check if we can merge it today.

**Review by peterlimg — CHANGES_REQUESTED** [review]

**Comment by devyetii** [comment]
@peterlimg  Sorry for the late reply. I've added a reply to your comment in the same day but actually wasn't added and it's still "Pending" and I don't know why 

Kindly note that this function (the one I added in this PR) doesn't check if the public key and the wallet ID matches, otherwise it checks that the operational wallet ID of the provider (the ID generated by transforming the PK of the provider, be it a miner, sharder, blobber, validator or authorizer) and the delegate wallet given as input to the registration function of this provider doesn't match. If they match, an error should be returned.

File: code/go/0chain.net/smartcontract/common/wallet.go
Comment (dabasov) [inline code/go/0chain.net/smartcontract/common/wallet.go]: why do you decode? there should be method to get walletId from pub

File: code/go/0chain.net/smartcontract/storagesc/blobber.go
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/blobber.go]: remove

File: code/go/0chain.net/smartcontract/minersc/sharder.go
Comment (dabasov) [inline code/go/0chain.net/smartcontract/minersc/sharder.go]: put this check into ValidateWallet for now

**Review by dabasov — CHANGES_REQUESTED** [review]

**Review by dabasov — APPROVED** [review]
lgtm

File: code/go/0chain.net/smartcontract/common/wallet.go
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/common/wallet.go]: Let's move this function into `encryption` package, and name it as `GetClientIDFromPublickKey()`, comment with `//` instead of `/**/`.

File: code/go/0chain.net/smartcontract/common/wallet.go
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/common/wallet.go]: Rename to `ValidateDelegateWallet()` and add comments to describe what it does, we need to ensure it would not be misused by users, i.e thought the the function will return nil if the clientID and pubkey matches

**Review by peterlimg — CHANGES_REQUESTED** [review]

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 1218-check-delegate-wallets  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/core/encryption/helpers.go
Comment (peterlimg) [inline code/go/0chain.net/core/encryption/helpers.go]: `*PublickKey` -> `*PublicKey`

File: code/go/0chain.net/core/encryption/helpers.go
Comment (peterlimg) [inline code/go/0chain.net/core/encryption/helpers.go]: We can put the function in keys.go, it's not a helper function actually.

**Review by peterlimg — DISMISSED** [review]

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**
